src/models/room.ts

```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
             return "Empty room";
         }
     }
+
+    /**
+     * When we receive a new visibility change event:
+     *
+     * - store this visibility change alongside the timeline, in case we
+     *   later need to apply it to an event that we haven't received yet;
+     * - if we have already received the event whose visibility has changed,
+     *   patch it to reflect the visibility change and inform listeners.
+     */
+    private applyNewVisibilityChangeEvent(event: MatrixEvent): void {
+        // Perform sanity checks on the event.
+        if (!event.isEventVisibilityChange()) {
+            // Internal error.
+            throw new Error("Expected a visibility change relation");
+        }
+
+        const relation = event.getRelation();
+        const originalEventId = relation.event_id;
+        let visible: boolean;
+        switch (event.getEventVisibilityChange()) {
+            case "hidden":
+                visible = false;
+                break;
+            case "visible":
+                visible = true;
+                break;
+            default:
+                // Event is ill-formed.
+                return;
+        }
```

Comment:
This can be shortened:

```suggestion
        visible = event.getEventVisibilityChange();
        if (!visible && visible !== false) return; // event is ill-formed
```

Reply:
Not sure I understand how you go from a string to a boolean.

Reply:
per internal: I'm not really seeing the value of the string, so would change this to be a boolean.